File: lib/services/expression_classifier.dart
import 'dart:math' as math;

import 'package:kwon_mediapipe_landmarker/kwon_mediapipe_landmarker.dart';

import '../models/expression.dart';

/// 基于 ARKit 52 blendshapes 的表情分类器。
///
/// 参考 Apple ARFaceAnchor.blendShapes 的语义，将若干 blendshape 系数
/// 组合为 7 种常见表情的打分，取最高者；若所有表情得分都很低则判为「中性」。
///
/// 这是规则式分类（非训练模型），便于调试阶段直观验证关键点 → 表情的映射。
class ExpressionClassifier {
  /// 判定为非中性表情所需的最小得分。
  final double activationThreshold;

  const ExpressionClassifier({this.activationThreshold = 0.28});

  ExpressionResult classify(Map<String, double> b) {
    if (b.isEmpty) return ExpressionResult.neutral;

    double v(String k) => (b[k] ?? 0).clamp(0.0, 1.0);
    double avg(String a, String c) => (v(a) + v(c)) / 2.0;

    // 各类基础特征
    final smile = avg(FaceBlendshape.mouthSmileLeft,
        FaceBlendshape.mouthSmileRight);
    final cheekSquint = avg(
        FaceBlendshape.cheekSquintLeft, FaceBlendshape.cheekSquintRight);
    final frown =
        avg(FaceBlendshape.mouthFrownLeft, FaceBlendshape.mouthFrownRight);
    final browInnerUp = v(FaceBlendshape.browInnerUp);
    final browOuterUp = avg(
        FaceBlendshape.browOuterUpLeft, FaceBlendshape.browOuterUpRight);
    final browDown =
        avg(FaceBlendshape.browDownLeft, FaceBlendshape.browDownRight);
    final jawOpen = v(FaceBlendshape.jawOpen);
    final eyeWide = avg(FaceBlendshape.eyeWideLeft, FaceBlendshape.eyeWideRight);
    final eyeSquint =
        avg(FaceBlendshape.eyeSquintLeft, FaceBlendshape.eyeSquintRight);
    final noseSneer =
        avg(FaceBlendshape.noseSneerLeft, FaceBlendshape.noseSneerRight);
    final upperLipUp = avg(
        FaceBlendshape.mouthUpperUpLeft, FaceBlendshape.mouthUpperUpRight);
    final mouthStretch = avg(
        FaceBlendshape.mouthStretchLeft, FaceBlendshape.mouthStretchRight);
    final mouthPress =
        avg(FaceBlendshape.mouthPressLeft, FaceBlendshape.mouthPressRight);

    final scores = <Expression, double>{
      // 高兴：嘴角上扬 + 脸颊上提（杜乡微笑）
      Expression.happy: smile * 0.8 + cheekSquint * 0.2,

      // 伤心：嘴角下垂 + 眉心内侧上抬（八字眉）
      Expression.sad: frown * 0.6 + browInnerUp * 0.4 - smile * 0.3,

      // 惊讶：张嘴 + 眉毛整体上抬 + 睁大眼
      Expression.surprised:
          jawOpen * 0.45 + browOuterUp * 0.25 + browInnerUp * 0.15 +
              eyeWide * 0.15,

      // 愤怒：皱眉（眉下压）+ 抿嘴/眯眼
      Expression.angry:
          browDown * 0.6 + mouthPress * 0.2 + eyeSquint * 0.2 - jawOpen * 0.2,

      // 厌恶：皱鼻 + 上唇上提
      Expression.disgusted: noseSneer * 0.6 + upperLipUp * 0.4,

      // 恐惧：睁大眼 + 眉内抬 + 嘴角横向拉伸
      Expression.fearful:
          eyeWide * 0.4 + browInnerUp * 0.3 + mouthStretch * 0.3,
    };

    Expression best = Expression.neutral;
    double bestScore = 0;
    scores.forEach((expr, s) {
      final clamped = s.clamp(0.0, 1.0);
      if (clamped > bestScore) {
        bestScore = clamped.toDouble();
        best = expr;
      }
    });

    final fullScores = <Expression, double>{
      Expression.neutral: math.max(0.0, 1.0 - bestScore),
      ...scores.map((k, val) => MapEntry(k, val.clamp(0.0, 1.0).toDouble())),
    };

    if (bestScore < activationThreshold) {
      return ExpressionResult(
        expression: Expression.neutral,
        score: fullScores[Expression.neutral] ?? 1.0,
        scores: fullScores,
      );
    }

    return ExpressionResult(
      expression: best,
      score: bestScore,
      scores: fullScores,
    );
  }
}

File: xbot-android/app/src/main/java/com/xbot/android/vision/ExpressionClassifier.kt
package com.xbot.android.vision

import com.xbot.android.model.Expression
import com.xbot.android.model.ExpressionResult
import kotlin.math.max

/**
 * 基于 ARKit 52 blendshapes 的规则式表情分类器（对应 Flutter ExpressionClassifier）。
 *
 * 参照 Apple ARFaceAnchor.blendShapes 的语义，将若干 blendshape 系数组合为 7 种常见
 * 表情的打分，取最高者；若所有表情得分都很低（< [activationThreshold]）则判为「中性」。
 *
 * 规则式分类（非训练模型），便于调试阶段直观验证关键点 → 表情的映射。
 *
 * blendshape key 为 MediaPipe tasks-vision 输出的 camelCase category name
 *（如 "eyeBlinkLeft", "mouthSmileLeft", "jawOpen", ...），与 ARKit 命名一致。
 */
class ExpressionClassifier(
    /** 判定为非中性表情所需的最小得分。 */
    private val activationThreshold: Float = 0.28f,
) {
    fun classify(b: Map<String, Float>): ExpressionResult {
        if (b.isEmpty()) return ExpressionResult.NEUTRAL

        fun v(k: String): Float = (b[k] ?: 0f).coerceIn(0f, 1f)
        fun avg(a: String, c: String): Float = (v(a) + v(c)) / 2f

        val smile = avg("mouthSmileLeft", "mouthSmileRight")
        val cheekSquint = avg("cheekSquintLeft", "cheekSquintRight")
        val frown = avg("mouthFrownLeft", "mouthFrownRight")
        val browInnerUp = v("browInnerUp")
        val browOuterUp = avg("browOuterUpLeft", "browOuterUpRight")
        val browDown = avg("browDownLeft", "browDownRight")
        val jawOpen = v("jawOpen")
        val eyeWide = avg("eyeWideLeft", "eyeWideRight")
        val eyeSquint = avg("eyeSquintLeft", "eyeSquintRight")
        val noseSneer = avg("noseSneerLeft", "noseSneerRight")
        val upperLipUp = avg("mouthUpperUpLeft", "mouthUpperUpRight")
        val mouthStretch = avg("mouthStretchLeft", "mouthStretchRight")
        val mouthPress = avg("mouthPressLeft", "mouthPressRight")

        val scores = LinkedHashMap<Expression, Float>()
        // 高兴：嘴角上扬 + 脸颊上提（杜乡微笑）
        scores[Expression.HAPPY] = smile * 0.8f + cheekSquint * 0.2f
        // 伤心：嘴角下垂 + 眉心内侧上抬（八字眉）
        scores[Expression.SAD] = frown * 0.6f + browInnerUp * 0.4f - smile * 0.3f
        // 惊讶：张嘴 + 眉毛整体上抬 + 睁大眼
        scores[Expression.SURPRISED] =
            jawOpen * 0.45f + browOuterUp * 0.25f + browInnerUp * 0.15f + eyeWide * 0.15f
        // 愤怒：皱眉（眉下压）+ 抿嘴/眯眼
        scores[Expression.ANGRY] =
            browDown * 0.6f + mouthPress * 0.2f + eyeSquint * 0.2f - jawOpen * 0.2f
        // 厌恶：皱鼻 + 上唇上提
        scores[Expression.DISGUSTED] = noseSneer * 0.6f + upperLipUp * 0.4f
        // 恐惧：睁大眼 + 眉内抬 + 嘴角横向拉伸
        scores[Expression.FEARFUL] = eyeWide * 0.4f + browInnerUp * 0.3f + mouthStretch * 0.3f

        var best = Expression.NEUTRAL
        var bestScore = 0f
        for ((expr, s) in scores) {
            val clamped = s.coerceIn(0f, 1f)
            if (clamped > bestScore) {
                bestScore = clamped
                best = expr
            }
        }

        val fullScores = LinkedHashMap<Expression, Float>()
        fullScores[Expression.NEUTRAL] = max(0f, 1f - bestScore)
        for ((k, value) in scores) fullScores[k] = value.coerceIn(0f, 1f)

        return if (bestScore < activationThreshold) {
            ExpressionResult(
                expression = Expression.NEUTRAL,
                score = fullScores[Expression.NEUTRAL] ?: 1f,
                scores = fullScores,
            )
        } else {
            ExpressionResult(expression = best, score = bestScore, scores = fullScores)
        }
    }
}
